File: typescript/ch5/Left.ts
import { Either } from './Either';
import { WrapperFunc } from './WrapperFunc';

class Left extends Either {
  get value() {
    throw TypeError(`Left(${this._value}) 값을 가져올 수 없습니다.`);
  }

  map(f: WrapperFunc) {
    return this;
  }

  getOrElse(other?: any) {
    return other;
  }

  orElse(f: WrapperFunc) {
    return f(this._value);
  }

  chain(f: WrapperFunc) {
    return this;
  }

  getOrElseThrow(value: any) {
    throw new Error(value);
  }

  filter(f: WrapperFunc) {
    return this;
  }

  toString() {
    return `Either.Left(${this._value})`;
  }
}

export {
  Left
};

File: python/python/ch5/Left.py
from .Either import Either

class Left(Either):
    @property
    def value(self):
        raise TypeError(f'Left({self._value}) 값을 가져올 수 없습니다.')
      
    def map(self, f=None):
        return self
      
    def get_or_else(self, other):
        return other
      
    def or_else(self, f):
        return f(self._value)
      
    def chain(self, f=None):
        return self
      
    def get_or_else_throw(self, value):
        raise ValueError(value)
      
    def filter(self, f=None):
        return self
      
    def __str__(self):
        return f'Either.Left({self._value})'